File: packages/flutter_tools/gradle/src/main/java/FlutterAppPluginLoaderPlugin.java
import org.gradle.api.GradleException;
import org.gradle.api.Plugin;
import org.gradle.api.initialization.Settings;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

class FlutterAppPluginLoaderPlugin implements Plugin<Settings> {
//    @Override
//    public void apply(Settings settings) {
//        File flutterProjectRoot = settings.getSettingsDir().getParentFile();
//
//        if(!settings.ext.hasProperty('flutterSdkPath')) {
//            def properties = new Properties()
//            def localPropertiesFile = new File(settings.rootProject.projectDir, "local.properties")
//            localPropertiesFile.withInputStream { properties.load(it) }
//            settings.ext.flutterSdkPath = properties.getProperty("flutter.sdk")
//            assert settings.ext.flutterSdkPath != null, "flutter.sdk not set in local.properties"
//        }
//
//        // Load shared gradle functions
//        settings.apply from: Paths.get(settings.ext.flutterSdkPath, "packages", "flutter_tools", "gradle", "src", "main", "groovy", "native_plugin_loader.groovy")
//
//        List<Map<String, Object>> nativePlugins = settings.ext.nativePluginLoader.getPlugins(flutterProjectRoot)
//        nativePlugins.each { androidPlugin ->
//                def pluginDirectory = new File(androidPlugin.path as String, 'android')
//            assert pluginDirectory.exists()
//            settings.include(":${androidPlugin.name}")
//            settings.project(":${androidPlugin.name}").projectDir = pluginDirectory
//        }
//    }
    @Override
    public void apply(Settings settings) {
        File flutterProjectRoot = settings.getSettingsDir().getParentFile();

        // Retrieve Flutter SDK path
        if (!settings.getExtensions().getExtraProperties().has("flutterSdkPath")) {
            Properties properties = new Properties();
            File localPropertiesFile = new File(settings.getRootProject().getProjectDir(), "local.properties");
            try (FileInputStream inputStream = new FileInputStream(localPropertiesFile)) {
                properties.load(inputStream);
                String flutterSdkPath = properties.getProperty("flutter.sdk");
                settings.getExtensions().add("flutterSdkPath", flutterSdkPath);
                if (flutterSdkPath == null) {
                    throw new GradleException("flutter.sdk not set in local.properties");
                }
            } catch (IOException e) {
                // Handle IOException appropriately
                e.printStackTrace();
            }
        }

        // Load shared Gradle functions
        String flutterSdkPath = (String) settings.getExtensions().getByName("flutterSdkPath");
        // Construct the path to 'native_plugin_loader.groovy' using Path
        Path nativePluginLoaderPath = Paths.get(flutterSdkPath, "packages", "flutter_tools", "gradle", "src", "main", "groovy", "native_plugin_loader.groovy");
        Map<String, String> options = new HashMap<>();
        options.put("from", nativePluginLoaderPath.toString());
        settings.apply(options);

        // Process native plugins
        NativePluginLoader nativePluginLoader = (NativePluginLoader) settings.getExtensions().getByName("nativePluginLoader");
        List<Map<String, Object>> nativePlugins = nativePluginLoader.getPlugins(flutterProjectRoot);
        for (Map<String, Object> androidPlugin : nativePlugins) {
            File pluginDirectory = new File((String) androidPlugin.get("path"), "android");
            assert pluginDirectory.exists();
            settings.include(":" + androidPlugin.get("name"));
            settings.project(":" + androidPlugin.get("name")).setProjectDir(pluginDirectory);
        }
    }

}

File: packages/flutter_tools/gradle/src/main/kotlin/FlutterAppPluginLoaderPlugin.kt
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package com.flutter.gradle

import org.gradle.api.Plugin
import org.gradle.api.initialization.Settings
import org.jetbrains.kotlin.gradle.plugin.extraProperties
import java.io.File
import java.nio.file.Paths
import java.util.Properties

private const val FLUTTER_SDK_PATH = "flutterSdkPath"

// TODO(gmackall): Migrate the settings plugin to a proper Gradle plugin, and out of the
//  buildscript block. This is a prerequisite for supporting the configuration cache.
@Suppress("unused") // This class is used by packages/flutter_tools/gradle/build.gradle.kts.
class FlutterAppPluginLoaderPlugin : Plugin<Settings> {
    override fun apply(settings: Settings) {
        val flutterProjectRoot: File = settings.settingsDir.parentFile

        if (!settings.extraProperties.has(FLUTTER_SDK_PATH)) {
            val localPropertiesFile = File(settings.rootProject.projectDir, "local.properties")
            if (localPropertiesFile.exists()) {
                val properties = Properties()
                localPropertiesFile.inputStream().use { properties.load(it) }
                properties.getProperty("flutter.sdk")?.let {
                    settings.extraProperties.set(FLUTTER_SDK_PATH, it)
                }
            }
            if (!settings.extraProperties.has(FLUTTER_SDK_PATH)) {
                val fallback =
                    System.getenv("FLUTTER_ROOT")
                        ?: settings.providers.gradleProperty("flutter.sdk").orNull
                if (fallback != null) {
                    settings.extraProperties.set(FLUTTER_SDK_PATH, fallback)
                }
            }
            check(settings.extraProperties.has(FLUTTER_SDK_PATH)) {
                "flutter.sdk not set in local.properties, the FLUTTER_ROOT environment variable, " +
                    "or the flutter.sdk Gradle property."
            }
        }

        settings.apply {
            from(
                Paths.get(
                    settings.extraProperties.get(FLUTTER_SDK_PATH) as String,
                    "packages",
                    "flutter_tools",
                    "gradle",
                    "src",
                    "main",
                    "scripts",
                    "native_plugin_loader.gradle.kts"
                )
            )
        }

        // The Flutter tool decides which plugins are consumed as prebuilt AARs and which are built
        // from source, and writes that decision out before invoking Gradle. When there is no plan
        // — Gradle invoked directly, or an SDK newer than this plugin understands — every plugin
        // is included as a subproject, which is the legacy behavior.
        val plan: FlutterPluginAarPlan? = FlutterPluginAarPlan.readOrNull(flutterProjectRoot)
        if (plan == null) {
            includeAllPluginsAsSubprojects(settings, flutterProjectRoot)
            return
        }

        plan.subprojectPlugins.forEach { plugin ->
            val pluginDirectory = File(plugin.path, "android")
            check(
                pluginDirectory.exists()
            ) { "Plugin directory does not exist: ${pluginDirectory.absolutePath}" }
            settings.include(":${plugin.name}")
            settings.project(":${plugin.name}").projectDir = pluginDirectory
        }
    }

    /**
     * Includes every Android plugin as a Gradle subproject, the model used before plugins could be
     * built as isolated AARs.
     */
    private fun includeAllPluginsAsSubprojects(
        settings: Settings,
        flutterProjectRoot: File
    ) {
        NativePluginLoaderReflectionBridge
            .getPlugins(settings.extraProperties, flutterProjectRoot)
            .forEach { androidPlugin ->
                val pluginDirectory = File(androidPlugin["path"] as String, "android")
                check(
                    pluginDirectory.exists()
                ) { "Plugin directory does not exist: ${pluginDirectory.absolutePath}" }
                val pluginName = androidPlugin["name"] as String
                settings.include(":$pluginName")
                settings.project(":$pluginName").projectDir = pluginDirectory
            }
    }
}
